pints/interfaces/_stan.py

```diff
@@ -0,0 +1,179 @@
+#
+# This file is part of PINTS (https://github.com/pints-team/pints/) which is
+# released under the BSD 3-clause license. See accompanying LICENSE.md for
+# copyright notice and full license details.
+#
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+import numpy as np
+from collections import Counter
+import pystan
+import pints
+
+
+class StanLogPDF(pints.LogPDF):
+    def __init__(self, stan_code, stan_data=None):
+        """
+        Creates a `pints.LogPDF` object from Stan code and data, which can
+        then be used in sampling, optimisation etc. Note, that this command
+        uses Pystan to interface with Stan which then goes on to compile the
+        underlying Stan model (see [1]_), so can take some time (typically
+        minutes or so) to execute.
+
+        If `pickle_filename` is provided, the object is pickled and can be
+        used to reload it later without recompiling the Stan model.
+
+        Note that the interface assumes that the parameters are on the
+        unconstrained scale (according to Stan's "constraint transforms" [1]_).
+        So, for example, if a variable is declared to have a lower bound of
+        zero, sampling happens on the log-transformed space. The interface
+        takes care of Jacobian transformations, so a user only needs to
+        transform the variable back to the constrained space (in the example,
+        using a `exp` transform) to obtain appropriate samples.
+
+        Extends :class:`pints.LogPDF`.
+
+        Parameters
+        ----------
+        stan_code
+            Stan code describing the model.
+        stan_data
+            Data in Python dictionary format as required by PyStan. Defaults to
+            None in which case `update_data` must be called to create a valid
+            Stan model fit object before calling.
+
+        References
+        ----------
+        .. [1] "Stan: a probabilistic programming language".
+               B Carpenter et al., (2017), Journal of Statistical Software
+        """
+
+        # compile stan
+        sm = pystan.StanModel(model_code=stan_code)
+        self._compiled_stan = sm
+
+        # only create stanfit if data supplied
+        if stan_data is None:
+            stanfit = None
+        else:
+            stanfit = sm.sampling(data=stan_data, iter=1, chains=1,
+                                  verbose=False, refresh=0,
+                                  control={'adapt_engaged': False})
+        self._initialise(stanfit)
+
+    def __call__(self, x):
+        if self._fit is None:
+            raise RuntimeError(
+                "No data supplied to create Stan model fit object. " +
+                "Run `update_data` first.")
+        vals = self._prepare_values(x)
+        try:
+            return self._log_prob(vals, adjust_transform=True)
+        # if Pints proposes a value outside of Stan's parameter bounds
+        except (RuntimeError, ValueError):
```

Comment:
Should we have some option to run without this error checking? Will a user be able to debug their stan code if we catch the errors here?

Reply:
Or raise a warning anytime this happens?

Reply:
I've added a warning to both times we use the try/except

Reply:
Is it useful to add the pystan error message?
```
except (...) as e:
  warnings.warn('Text: ' + str(e))
```

Reply:
Great idea! Done.